src/charm.py

```diff
@@ -561,29 +575,75 @@ def _on_mongod_pebble_ready(self, event) -> None:
 
         except (PathError, ProtocolError, MissingSecretError) as e:
             logger.error("Cannot initialize workload: %r", e)
-            event.defer()
-            return
+            return False
 
-        # Add initial Pebble config layer using the Pebble API
-        container.add_layer("mongod", self._mongod_layer, combine=True)
-        container.add_layer("log_rotate", self._log_rotate_layer, combine=True)
+        return True
+
+    def __configure_layers(self, container: Container) -> bool:
+        """Configure the layers of the container."""
+        modified = False
+        current_layers = container.get_plan()
+        new_layers = {
+            Config.SERVICE_NAME: self._mongod_layer,
+            "log_rotate": self._log_rotate_layer,
+        }
         if self.is_role(Config.Role.CONFIG_SERVER):
-            container.add_layer("mongos", self._mongos_layer, combine=True)
+            new_layers["mongos"] = self._mongos_layer
 
+        # Add Pebble config layers missing or modified
+        for layer_name, definition in new_layers.items():
+            for service_name, service in definition.services.items():
+                if current_layers.services.get(service_name) != service:
+                    modified = True
+                    logger.debug(f"Adding layer {service_name}.")
+                    container.add_layer(layer_name, definition, combine=True)
+
+        # We'll always have a logrotate configuration at this point.
         container.exec(["chmod", "644", "/etc/logrotate.d/mongodb"])
 
-        # Restart changed services and start startup-enabled services.
-        container.replan()
+        return modified
+
+    def _configure_container(self, container: Container) -> bool:
```

Comment:
I can raise instead of returning a boolean indicating success/failure but it will be the same same from the caller side.
It's just that since I exported part of the code that would normally make the caller exit in case of failure, I need to propagate this information to the caller.

Reply:
> I can raise instead of returning a boolean indicating success/failure but it will be the same same from the caller side.

yes please, this would be great 🙏

Reply:
It's been done.